src/components/Personalization/helper/index.js

```diff
@@ -24,23 +26,34 @@ export { default as setAttributes } from "./setAttributes";
 export { default as swapImage } from "./swapImage";
 export { default as rearrangeChildren } from "./rearrangeChildren";
 
+const renderContent = (elements, content, renderFunc) => {
+  const executions = elements.map(element =>
+    Promise.resolve(renderFunc(element, content))
+  );
+
+  return Promise.all(executions);
+};
+
 export const createAction = (collect, renderFunc) => {
-  return (settings, event) => {
-    const { elements, prehidingSelector } = event;
-    const { content, meta } = settings;
-    const executions = elements.map(element =>
-      Promise.resolve(renderFunc(element, content))
-    );
-
-    return Promise.all(executions)
-      .then(() => {
-        // Success, unhide elements and notify
-        showElements(prehidingSelector);
-        collect(meta);
-      })
-      .catch(() => {
-        // Something went horribly wrong, unhide elements
-        showElements(prehidingSelector);
-      });
+  return settings => {
+    const { selector, prehidingSelector, content, meta } = settings;
+
+    hideElements(prehidingSelector);
+
+    return awaitSelector(selector, selectNodesWithEq)
+      .then(elements => renderContent(elements, content, renderFunc))
+      .then(
+        () => {
+          // if everything is OK, notify and show elements
+          collect(meta);
+          showElements(prehidingSelector);
+        },
+        () => {
+          // in case of awaiting timing or error, we need to remove the style tag
+          // hence showing the pre-hidden elements and notify
+          collect(meta);
```

Comment:
Will this `reject` callback catch all the exceptions in this flow, or should keep those in a `catch`?

Reply:
@jfkhoury I have tried different combinations and the `rejected` callback catches any error that might be thrown.

An alternative approach would be to guard `collect()` and `showElements()` using a `try/catch` and just have `then().catch()`.

We don't want to accidentally invoke collect twice, although I think it would be good to augment the `error` collect call with an error object, so we could report back to conductor that something is broken.

Reply:
This looks good to me.  I like the idea of augmenting the error collect call.

Reply:
I'll leave as is and I'll add a separate ticket to capture errors in `collect` call.